server/src/main/java/org/elasticsearch/action/admin/indices/stats/IndicesStatsResponse.java

```diff
@@ -83,26 +82,18 @@ public IndexStats getIndex(String index) {
         if (indicesStats != null) {
             return indicesStats;
         }
-        Map<String, IndexStats> indicesStats = new HashMap<>();
 
-        Set<Index> indices = new HashSet<>();
+        final Map<String, IndexStatsBuilder> indicesStatsBuilder = new HashMap<>();
         for (ShardStats shard : shards) {
-            indices.add(shard.getShardRouting().index());
-        }
-
-        for (Index index : indices) {
-            List<ShardStats> shards = new ArrayList<>();
+            Index index = shard.getShardRouting().index();
             String indexName = index.getName();
-            for (ShardStats shard : this.shards) {
-                if (shard.getShardRouting().getIndexName().equals(indexName)) {
-                    shards.add(shard);
-                }
-            }
-            indicesStats.put(
-                indexName, new IndexStats(indexName, index.getUUID(), shards.toArray(new ShardStats[shards.size()]))
-            );
+            IndexStatsBuilder indicesStatsBuildr = indicesStatsBuilder.computeIfAbsent(indexName,
```

Comment:
I find this naming misleading. `indicesStatsBuildr` looks awfully similar to the HashMap `indicesStatsBuilder`, but without an `e`. I think the HashMap could use a different name that reflects its data-structure. what do you think?